lib/gate_appl.h

```diff
@@ -62,50 +63,150 @@ inline void CalcMatrix4(unsigned q0, unsigned q1,
+ * @param state The state of the system, to be updated by this method.
+ * @return True if the measurement performed successfully; false otherwise.
+ */
+template <typename Simulator, typename Gate, typename Rgen>
+inline bool ApplyGate(const typename Simulator::StateSpace& state_space,
+                      const Simulator& simulator, const Gate& gate, Rgen& rgen,
+                      typename Simulator::State& state) {
+  using MeasurementResult = typename Simulator::StateSpace::MeasurementResult;
+  std::vector<MeasurementResult> discarded_results;
+  return
+      ApplyGate(state_space, simulator, gate, rgen, state, discarded_results);
+}
+
+/**
+ * Applies the given fused gate to the simulator state. Ignores measurement
+ *   gates.
+ * @param simulator Simulator object. Provides specific implementations for
+ *   applying gates.
+ * @param gate The gate to be applied.
+ * @param state The state of the system, to be updated by this method.
+ */
+template <typename Simulator, typename Gate>
+inline void ApplyFusedGate(const Simulator& simulator, const Gate& gate,
+                           typename Simulator::State& state) {
+  if (gate.kind != gate::kMeasurement) {
+    typename Simulator::fp_type matrix[32];
+
+    if (gate.num_qubits == 1) {
+      CalcMatrix2(gate.gates, matrix);
+    } else if (gate.num_qubits == 2) {
+      CalcMatrix4(gate.qubits[0], gate.qubits[1], gate.gates, matrix);
+    }
+
+    simulator.ApplyGate(gate.qubits, matrix, state);
+  }
+}
+
+/**
+ * Applies the given fused gate to the simulator state.
+ * @param state_space StateSpace object required to perform measurements.
+ * @param simulator Simulator object. Provides specific implementations for
+ *   applying gates.
+ * @param gate The gate to be applied.
+ * @param rgen Random number generator to perform measurements.
+ * @param state The state of the system, to be updated by this method.
+ * @param mresults As an input parameter, this can be empty or this can
+ *   contain the results of the previous measurements. If gate is a measurement
+ *   gate then after a successful run, the measurement result will be added to
+ *   this.
+ * @return True if the measurement performed successfully; false otherwise.
+ */
+template <typename Simulator, typename Gate, typename Rgen>
+inline bool ApplyFusedGate(
+    const typename Simulator::StateSpace& state_space,
+    const Simulator& simulator, const Gate& gate, Rgen& rgen,
+    typename Simulator::State& state,
+    std::vector<typename Simulator::StateSpace::MeasurementResult>& mresults) {
+  if (gate.kind == gate::kMeasurement) {
+    auto measure_result = state_space.Measure(gate.qubits, rgen, state);
+    if (measure_result.valid) {
+      mresults.push_back(std::move(measure_result));
+    } else {
+      return false;
+    }
+  } else {
+    ApplyFusedGate(simulator, gate, state);
+  }
+
+  return true;
+}
+
+/**
+ * Applies the given fused gate to the simulator state.
```

Comment:
This comment should explicitly note that the function discards measurement results.

Reply:
Done.